mlflow/langchain/_langchain_autolog.py

```diff
@@ -77,34 +77,45 @@ def patched_inference(func_name, original, self, *args, **kwargs):
 
     We patch inference functions for different models based on their usage.
     """
-    # Inject MLflow tracer into the model
-    from mlflow.langchain.langchain_tracer import MlflowLangchainTracer
-
-    config = AutoLoggingConfig.init()
-    if config.log_traces:
-        mlflow_tracer = MlflowLangchainTracer()
-        args, kwargs = _inject_mlflow_callbacks(func_name, [mlflow_tracer], args, kwargs)
-        _logger.debug("Injected MLflow callbacks into the model.")
 
     # NB: Running the original inference with disabling autologging, so we only patch the top-level
     # component and avoid duplicate logging for child components.
     def _invoke(self, *args, **kwargs):
         with disable_autologging():
             return original(self, *args, **kwargs)
 
-    # Traces does not require an MLflow run, only the other optional artifacts require it.
-    if config.should_log_optional_artifacts():
-        with _setup_autolog_run(config, self) as run_id:
+    config = AutoLoggingConfig.init()
+    with patch_args_and_kwargs(func_name, config, args, kwargs) as (args, kwargs):
+        # Traces does not require an MLflow run, only the other optional artifacts require it.
+        if config.should_log_optional_artifacts():
+            with _setup_autolog_run(config, self) as run_id:
+                result = _invoke(self, *args, **kwargs)
+            _log_optional_artifacts(config, run_id, result, self, func_name, *args, **kwargs)
+        else:
             result = _invoke(self, *args, **kwargs)
-        _log_optional_artifacts(config, run_id, result, self, func_name, *args, **kwargs)
-    else:
-        result = _invoke(self, *args, **kwargs)
-
-    if config.log_traces:
-        mlflow_tracer.flush_tracker()
     return result
 
 
+@contextlib.contextmanager
+def patch_args_and_kwargs(func_name, autologging_config, args, kwargs):
+    from mlflow.langchain.langchain_tracer import MlflowLangchainTracer
+
+    if autologging_config.log_traces:
+        # Inject MLflow tracer into the model
```

Comment:
```suggestion
        # Inject MLflow tracer into the callbacks argument
```

just to be more accurate